modules/gapi/test/gapi_async_test.cpp

```diff
@@ -380,14 +380,117 @@ TYPED_TEST_P(cancel, basic){
+                case GRunArg::index_of<cv::Scalar>()              :   result.emplace_back(&util::get<cv::Scalar>(arg));  break;
+                case GRunArg::index_of<cv::UMat>()                :   result.emplace_back(&util::get<cv::UMat>(arg));    break;
+    #endif // !defined(GAPI_STANDALONE)
+                case GRunArg::index_of<cv::gapi::own::Mat>()      :   result.emplace_back(&util::get<cv::gapi::own::Mat>   (arg));   break;
+                case GRunArg::index_of<cv::gapi::own::Scalar>()   :   result.emplace_back(&util::get<cv::gapi::own::Scalar>(arg));   break;
+                case GRunArg::index_of<cv::detail::VectorRef>()   :   result.emplace_back(util::get<cv::detail::VectorRef>  (arg));   break;
+                default : ;
+            }
+        }
+        return result;
+    }
+}
+
 REGISTER_TYPED_TEST_CASE_P(cancel, basic);
 
+template<typename case_t>
+struct output_args_lifetime : ::testing::Test{
+    static constexpr const int num_of_requests = 20;
+};
+TYPED_TEST_CASE_P(output_args_lifetime);
+//There are intentionaly no actual checks (asserts and verify) in output_args_lifetime tests.
+//They are more of example use-cases nature than tests. (ASAN/valgrind can steal catch issues here)
+TYPED_TEST_P(output_args_lifetime, callback){
+
+    std::atomic<int> active_requests = {0};
+
+    for (int i=0; i<this->num_of_requests; i++)
+    {
+        TypeParam r;
+
+        //While input args are copied on on call to async function (actually on call to cv::gin)
+        //and thus do not have to outlive the actual completion of asynchronous activity.
+        //Output arguments are "captured" by reference and therefore _must_ outlive the asynchronous activity
+        //(i.e. live at least until callback is called)
+        auto out_args_ptr =  std::make_shared<cv::GRunArgs>(deep_copy_out_args(r.out_args()));
+
+        //Extend lifetime of out_mat_ptr content by capturing it into a callback
+        auto cb =  [&active_requests, out_args_ptr](std::exception_ptr ){
+            --active_requests;
+        };
+
+        ++active_requests;
+
+        r.async(cb, r.in_args(), args_p_from_args(*out_args_ptr));
+    }
+
+
+   while(active_requests){
+       std::this_thread::sleep_for(std::chrono::milliseconds{2});
+   }
+}
+
+
+TYPED_TEST_P(output_args_lifetime, future){
+    auto const num_of_requests = this->num_of_requests;
+
+    std::vector<std::future<void>>                      fs(num_of_requests);
+    std::vector<std::shared_ptr<cv::GRunArgs>>    out_ptrs(num_of_requests);
+
+    for (int i=0; i<num_of_requests; i++)
+    {
+        TypeParam r;
+
+        //While input args are copied on on call to async function (actually on call to cv::gin)
+        //and thus do not have to outlive the actual completion of asynchronous activity.
+        //Output arguments are "captured" by reference and therefore _must_ outlive the asynchronous activity
+        //(i.e. live at least until future.get()/wait() is returned)
```

Comment:
I feel like this should be added to the API documentation as well. And when it's added there we can assume that user is acquainted with the logic, so a simpler comment would suffice here.

I'd rephrase, let's avoid technical details, the following should work, I suppose:
```
Note: 
Input args are __copied__ and output args are __captured by reference__
User __must__ ensure output arguments are "alive" and accessible from another thread when leaving the current scope
```

Reply:
same applies to the above comment in callback version

Reply:
changed

Reply:
good, consider resolved